src/RemediationRoleSaver.cpp

```diff
@@ -79,35 +151,194 @@ void RemediationSaverBase::saveToFile(const QString& filename)
     }
 }
 
-QString RemediationSaverBase::guessFilenameStem() const
+
+BashProfileRemediationSaver::BashProfileRemediationSaver(QWidget* parentWindow, ScanningSession* session):
+    ProfileBasedRemediationSaver(parentWindow, session,
+            bashSaveMessage, bashFiletypeExtension, bashFiletypeTemplate, bashFixTemplate)
+{}
+
+
+AnsibleProfileRemediationSaver::AnsibleProfileRemediationSaver(QWidget* parentWindow, ScanningSession* session):
+    ProfileBasedRemediationSaver(parentWindow, session,
+            ansibleSaveMessage, ansibleFiletypeExtension, ansibleFiletypeTemplate, ansibleFixType)
+{}
+
+
+PuppetProfileRemediationSaver::PuppetProfileRemediationSaver(QWidget* parentWindow, ScanningSession* session):
+    ProfileBasedRemediationSaver(parentWindow, session,
+            puppetSaveMessage, puppetFiletypeExtension, puppetFiletypeTemplate, puppetFixType)
+{}
+
+
+#ifndef SCAP_WORKBENCH_USE_LIBRARY_FOR_RESULT_BASED_REMEDIATION_ROLES_GENERATION
+ResultBasedProcessRemediationSaver::ResultBasedProcessRemediationSaver(QWidget* parentWindow, const QByteArray& arfContents,
+        const QString& saveMessage, const QString& filetypeExtension, const QString& filetypeTemplate, const QString& fixType):
+    RemediationSaverBase(parentWindow, saveMessage, filetypeExtension, filetypeTemplate, fixType)
 {
-    // TODO: Add guess that uses benchmark and profile names
-    return QString("remediation");
+    mArfFile.setAutoRemove(true);
+    mArfFile.open();
+    mArfFile.write(arfContents);
+    mArfFile.close();
 }
 
 
-BashRemediationSaver::BashRemediationSaver(QWidget* parentWindow, ScanningSession* session):RemediationSaverBase(parentWindow, session)
+void ResultBasedProcessRemediationSaver::saveToFile(const QString& filename)
 {
-    mSaveMessage = QObject::tr("Save remediation role as a bash script");
-    mFiletypeExtension = "sh";
-    mFiletypeTemplate = QObject::tr("bash script (*.%1)");
-    mFixTemplate = QObject::tr("urn:xccdf:fix:script:sh");
+    QStringList args;
+    args.append("xccdf");
+    args.append("generate");
+    args.append("fix");
+
+    args.append("--template");
+    args.append(mTemplateString);
+    args.append("--output");
+    args.append(filename);
+
+    // vvv This will work, if there is only one result ID in the ARF file, it will be picked no matter what the argument value is.
+    // However, ommitting --result-id "" won't work.
+    args.append("--result-id");
+    args.append("");
```

Comment:
Result-id should not be "". I wonder why it works, it looks like a hack.

Reply:
This is difficult to tell. The `--result-id` is in fact a result suffix, And an empty string is a suffix of every result ID, which is convenient as it is guaranteed that we have only one result per ARF.
I think that having this as part of the official `oscap` API is better than implementing the tedious result ID extraction here in the workbench.